src/coreclr/src/jit/optimizer.cpp

```diff
@@ -7657,17 +7673,37 @@ void Compiler::optComputeLoopNestSideEffects(unsigned lnum)
 {
     assert(optLoopTable[lnum].lpParent == BasicBlock::NOT_IN_LOOP); // Requires: lnum is outermost.
     BasicBlock* botNext = optLoopTable[lnum].lpBottom->bbNext;
+    JITDUMP("optComputeLoopSideEffects botNext is " FMT_BB ", lnum is %d\n", botNext->bbNum, lnum);
     for (BasicBlock* bbInLoop = optLoopTable[lnum].lpFirst; bbInLoop != botNext; bbInLoop = bbInLoop->bbNext)
     {
-        optComputeLoopSideEffectsOfBlock(bbInLoop);
+        if (!optComputeLoopSideEffectsOfBlock(bbInLoop))
```

Comment:
It seems that calling `optComputeLoopSideEffectsOfBlock()` will only result in `false` and taking this branch if `bbNatLoopNum == BasicBlock::NOT_IN_LOOP` (see `optComputeLoopSideEffectsOfBlock()` below). However, immediately below this check we also check `bbNatLoopNum != BasicBlock::NOT_IN_LOOP` as the predicate to a `while` loop, but the only way we got in here is if it was `BasicBlock::NOT_IN_LOOP`. So is the `while` loop needed?

Reply:
Yes,  Each block can be in a set of nested loops.
`lnum `start out being the inner most loop number and each time through this while it gets assigned the loop parent.  Thus this while loop sets `lpLoopHasMemoryHavoc `on all of the loops containing this block,
from innermost to outermost.
This code is actually copied & modified from line 7961 of `optimizer.cpp`

Reply:
But the `lnum` is a local below and a copy of the value that was checked in `optComputeLoopSideEffectsOfBlock()`? So if the check above already knows `bbNatLoopNum` is `BasicBlock::NOT_IN_LOOP` which it is assigned to the local then the check will never be true. Should the `lnum` be outside of the if check?

Reply:
Looking at 7961, I do think that `lnum` needs to be outside of the `for` loop.

Reply:
I will refactor this to be a function call

Reply:
Oh! This rework makes why this worked clearer, but I would suggest addressing something. The input argument is `lnum` and there is the inner local _also_ called `lnum` that is shadowing the input argument. You might want to change the inner name. That should be a compiler warning...

Reply:
Opps, I actually meant to delete that entire block (declaration of lnum and the loop) as it performs the same work as the call to `optRecordLoopNestsMemoryHavoc `that I added